### ISSUE TITLE
[Feature Request] Implement union types (rather than nominal enums)

### ISSUE THREAD

**Issue by nmsmith** [modularml/mojo#43]
### Summary

I propose to consider adding TypeScript-style union types to Mojo — instead of Rust-style nominal enums. This approach seems like it would work well given that Mojo plans to extend Python, which is a dynamically-typed language. In addition to enabling type declarations for Python libraries, union types would allow Mojo users to avoid the modelling issued associated with nominal enums.

### Motivation

Most statically-typed languages that implement type-safe enums use nominal typing. For example, in Rust, to define a function that can return two different error types, one must first define (and _name_) the error type:
```
enum IOParseError {
    IOError(...),
    ParseError(...),
}
```

One can then specify the return type of the error-producing function as follows:
```
fn parseFile(file: &str) -> Result<Success, IOParseError>
```

This approach doesn't scale when different functions can return different combinations of errors. For example, maybe we have:

- The function `parseFile` (above) that produces either an `IOError` or a `ParseError`.
-  A function `readFile` that only produces an `IOError`.
- A function `parseString` that only produces a `ParseError`.
- A function `parseAndExecuteFile` that produces either an `IOError`, a `ParseError`, or a `RuntimeError`.
- ...etc.

The issue here is that each combination of errors requires a different enum to be declared, and each enum needs to have _unique_ constructor names. In short: expressing these return types using nominal enums is a nightmare. (And this problem isn't unique to error types. There are many other scenarios in which this issue occurs.)

In contrast, modelling these functions in a dynamically-typed language (such as Python) is trivial. For example, the implementation of `parseFile` will simply return whatever error object is appropriate. No up-front declaration is required.

Because of this flexibility afforded by dynamically-typed languages, retroactively adding a type system to such languages has proved challenging. But in recent years, TypeScript has demonstrated that it _is_ possible to statically-type these kinds of programs. TypeScript's trick is to offer _union_ types, which can directly model a disjunction of types:
```
function parseFile():           Result<Success, IOError | ParseError> {...}
function readFile():            Result<Success, IOError> {...}
function parseString():         Result<Success, ParseError> {...}
function parseAndExecuteFile(): Result<Success, IOError | ParseError | RuntimeError> {...}
```

Thus, I see two major motivations for Mojo to offer TypeScript-style union types:
- For the modelling advantages discussed above.
- To allow Mojo users to write type declarations for Python libraries, just as TypeScript developers are able to write type declarations for JavaScript libraries.

### Design

Developing a precise semantics for union types will require a large amount of work, of course. Thankfully, TypeScript's semantics are a helpful piece of prior art.

Here are some issues that will require consideration:
- Is it possible to compile union types as efficiently as nominal enums?
    - I suspect that the answer is "yes", at least in cases where union types are used in the same manner as enums. For example, if one defines the union type `type Color = Red | Green | Blue` (where `Red`, `Green`, and `Blue` are pre-defined, empty `struct`s), there is no reason that the compiler can't encode them via the integers `0`, `1` and `2` when storing them in variables of type `Color` and when passing them as arguments of type `Color`. A difficulty arises when the same structs are used in _multiple_ enums, e.g. `type TrafficLight = Red | Yellow | Green`. When these two types "interact" (e.g. a `Color` value is used as a `TrafficLight` argument), _re-mapping_ of the tag values may need to occur, e.g. `Green` gets mapped from `1` to `2`. Alternatively, the compiler can avoid the need for such re-mappings by monomorphizing function calls. For example, given a function with an argument of type `TrafficLight`, the compiler could generate a "special" version of the function where the tag of each `TrafficLight` color is expected to match the tag for the respective `Color` color (i.e. `Green` is encoded as `1`). Call sites that provide a `Color` as an argument will invoke this version of the function.
- If the elements of a type union are `struct`s, rather than enum-style "named alternatives" (known as _data constructors_ in FP), how would pattern-matching and destructuring work?
    - The best approach is probably to pattern-match directly on `struct`s. No special "constructor" concept is required. However, how would that work? Not all structs are "plain old data": many `__init__` methods are not invertible. We would probably need a keyword or decorator to designate that a `struct` is destructurable. The `@value` decorator that Mojo already supports seems sufficient. Another alternative would be to allow structs to define their own destructuring behaviour — like how `__getitem__` works.  Structs that are _not_ destructurable can still be used in pattern-matching, as long as no attempt is made to destructure them.
- Union types are typically understood to work like _sets_, meaning that duplicate occurrences of a type are eliminated. For example, `None | None = None`. How should unification be treated?
     - As a case study, one might try to define `Optional[T]` as follows:
     `type Optional[T] = T | None`
     However, this definition suffers from the issue (?) that `Optional[None]` is just `None`, and that `Optional[Optional[Int]]` is just `Optional[Int]`. This seems hard to reason about!
     - The "nuclear option" to addressing this issue would be to disallow type declarations that are able to unify. So the above type would be disallowed, and users would have to write a definition like `type Optional[T] = Some[T] | None` instead (where `Some` is a pre-defined `struct`). It seems reasonable to impose this restriction, given that it still allows all traditional enumerations to be defined.

This is probably enough detail for the moment. What does the Mojo team think of this approach?

**Comment by lsh** [comment]
For what it's worth, the optional type [already on the roadmap](https://docs.modular.com/mojo/roadmap.html#small-independent-features).

**Comment by lattner** [comment]
Yes, this is a very important feature, also very important to Swift and many ML languages.  Doing this right is super important.  Thank you for filing this!

**Comment by nmsmith** [comment]
It's worth clarifying that my proposal is to _avoid_ the design that Swift, Rust, and ML have implemented, owing to its lack of _Modularity_. 🥁 

What are the next steps for this feature? Will Modular be sharing more thoughts in the future?

**Comment by nmn** [comment]
So looking at the original Issue and the responses a think there has been some confusion.

## Support for Structural Union Types

@nmsmith is asking for Structural Union Types *instead of* Nominal Algebraic types such as in Swift and Rust. (Optionals in Swift are an exception)

### The Good

This would be a big improvement in developer experience because you no longer need to “wrap” your types in enums when you are choosing one of multiple types.

### The Bad

Usually, supporting true structural unwrapped union types means more runtime overhead as you need the ability to determine the actual type at runtime. 

Nominal enums, on the other hand need minimal overhead (essentially, just a numberic “tag”) to determine which “case” it is.

### Both Unions and Algebraic Data Type Enums should probably exist in the language

Mojo is an interesting case since it aims to be a superset of Python. Since Python is a dynamic language, all types in it can be determined at runtime. A structural Union type makes sense for this.

On the other hand, Mojo also provides primitives such as `struct` for the maximum possible performance.

Keeping this in mind, structural types and union types make sense for the “Python, but faster case”. But when you want to eek out the best possible performance from your program, you’d probably want to move from Structural Union types to Nominal Enums.

**Comment by nmsmith** [comment]
@nmn Yes, you've summarized the distinction well. Thanks for that.

I addressed the topic of runtime overhead in my original post, under the following bullet point:
> Is it possible to compile union types as efficiently as nominal enums?

To summarize it: I believe that if you're using a structural union type in the same manner that you'd be using an enum, you can compile it in exactly the same way — i.e. efficiently, using numeric tags. (Why shouldn't that be the case? If the code is isomorphic to the code for enums, the compiler can use the same compilation strategy.)

If this is true, there would be no reason to support nominal enums in Mojo. They would be entirely redundant.

Also, it's worth demonstrating that translating any enum to a union type is trivial. Consider the following enum:
```
enum Foo {
    A(Int),
    B(Int, Int),
}
```

This can be re-written as a union type as follows. (I shall assume that our hypothetical programming language has a syntax for "tuple structs".)
```
struct A(Int)
struct B(Int, Int)

type Foo = A | B
```

Notably, this requires the same number of lines of code as the enum declaration. The advantage is, of course, that structs `A` and `B` can thereafter be used in _any number of unions_:

```
type Bar = A | B | C
type Qux = C | A
```

I gave an example of the utility of this in my original post. The type unions shown below are _incredibly practical_:
```
function parseFile():           Result<Success, IOError | ParseError> {...}
function readFile():            Result<Success, IOError> {...}
function parseString():         Result<Success, ParseError> {...}
function parseAndExecuteFile(): Result<Success, IOError | ParseError | RuntimeError> {...}
```

The only way to express these types signatures with enums is to add a lot of boilerplate. (Try it yourself!)

**Comment by lsh** [comment]
I agree that many error types can be tough, but I would also caution against giant catch all error types. Sabrina Jewson has a [great article](https://sabrinajewson.org/blog/errors) describing the pitfalls of multierror unions.

**Comment by nmn** [comment]
> I believe that if you're using a structural union type in the same manner that you'd be using an enum, you can compile it in exactly the same way — i.e. efficiently, using numeric tags.

This is not as simple as you suggest. But in order to not pollute this issue anymore I'm going to not go into detail here. TLDR; you'd need a unique tag for every type for what you're describing to be doable. You may argue that that is worth it, but it is overhead.

**Comment by nmsmith** [comment]
I'm assuming you mean each `struct` would need to have a globally-unique tag. In my original post I argued that you wouldn't need to do this. The tag that is associated with a particular member of a union can be chosen according to how the union is ultimately used throughout the program. If the union is used just like a normal enum — i.e. if you don't reuse those same structs in other unions — then they can be given the same tags that enums are given. And even if you _do_ have overlapping unions, they can still be compiled like enums, as long as a struct value isn't passed between pieces of code that tag things differently. If a struct _is_ used in that way, then the compiler would need to either re-map the tags at the point where a struct moves between unions, or monomorphize the program (the same trick used for generics) such that no remapping is necessary.

**Comment by nmn** [comment]
@nmsmith I know what you argued. I had written a full reply why that wouldn't be feasible but it got long so I decided against it.

You've generally already mentioned the problem I would describe. You'd need to know how the type is used during *all* execution. This is possible for small programs, but not when you start building libraries, partial re-compilation etc.

> or monomorphize the program (the same trick used for generics) such that no remapping is necessary.

There's a bunch of complexity you're papering over here as well, but let's not get into that!

**Comment by nmsmith** [comment]
There are some challenges, but I think they are surmountable. It's just another optimization task — compilers are full of those. (Traditional enums are already subjected to various kinds of optimization.)

**Comment by lsh** [comment]
I find some parts of the update compelling, but I have a couple questions and a nit.

First, when you say `drawThePage[T: Draw]` works because it has `List[Button | Image]`, that means you are implying that `Button | Image` implements `Draw`. The question of when to apply this implementation comes up. When one specifies `Button | Image`, does it have to go through, find every trait each variant contains, and implement the intersection? One problem with that approach is one might want to implement it differently for the union of the two compared to the individual.

Also, as a nit, your comparison with Rust is not exactly correct, and therefore not addressing trait objects. The Rust could would be more verbose, but would not be heap allocated. The comparison Rust code would have an enum variant like `ImageOrButton { Image(Image), Button(Button) }`, and implement `Draw` for the enum. Trait objects are explicitly for cases where one cannot know the concrete type at compile time.

A final question is about the rebinding. In Rust that would be considered a `downcast` and could be fallible. Part of the reason for this is that some APIs use traits as a way to make opaque types so that they can avoid leaking implementation details to users (such as returning `impl Iterator`). I don't see any discussion of that use case here.

**Comment by nmsmith** [comment]
> When one specifies Button | Image, does it have to go through, find every trait each variant contains, and implement the intersection?

The compiler only needs to check whether these types implement `Draw` if they are passed into a function/method that expects them to implement it. `drawThePage` expects this, so this is the point at which the compiler needs to check.

> One might want to implement [the Draw trait] differently for the union.

Under this proposal, that would not be possible. What's more, I cannot think of a case where that would make sense. The union operator `|` fundamentally represents the everyday concept of `or`. In everyday life, I can draw a bird, and I can draw a cat, but I cannot draw a "bird or cat". I have to pick _one_ of the options to draw, and each option corresponds to a different method of drawing.

> The comparison Rust code would have an enum variant like ImageOrButton { Image(Image), Button(Button) }.

This wouldn't work. In the implementation of `drawThePage`, the type of the list is widened to `[T | MenuBar | SideBar]`. You can't simulate this using enums. (Not without introducing a ton of boilerplate, at least.) 

> Trait objects are explicitly for cases where one cannot know the concrete type at compile time.

If Mojo is carefully designed, the compiler can _always_ know the concrete types at compile-time. (Assuming we are talking about a subset of Mojo where types cannot be constructed dynamically, as they can in Python.) For example, I could have defined the `drawThePage` function in a separate library, and the compilation strategy would be the same. Namely, `drawThePage` is specialized (monomorphized) to the particular value of its type parameter `T` that the caller provides. There is no way for a library to **hide the types from the compiler:** the compiler is able to see everything at the moment of specialization. (It's important to distinguish between the idea of hiding types from the programmer, versus hiding types from the compiler. The former is a useful notion, but to the best of my knowledge, the latter is not.)

That said, if we take a step back and consider the broader picture, we can identify a true utility that trait objects provide. Trait objects can be understood as a special case of _runtime generics_. Runtime generics are certain useful, because they allow functions to be compiled without needing to know which concrete parameters (the stuff in square brackets `[]`) they will be provided with. Mojo will probably support runtime generics at some point, but I imagine it will be a more general design.

Runtime generics would have utility in a hypothetical situation where separately-compiled Mojo programs are trying to communicate with one another, e.g. a client written by one Mojo user is trying to send data to a server written by another Mojo user. But that's a deep rabbit hole — we have to talk about stable ABIs and communication protocols!

Excluding that use-case, the trade-off between runtime generics and compile-time generics (i.e. templates) is ultimately about performance: code size vs. runtime overhead. Therefore, it's probably a mistake to offer runtime generics via a special language construct such as trait objects. Instead, it makes more sense to me to allow users to apply an annotation to a parameter to indicate that the function should remain generic with respect to that parameter, rather than specializing to it. For example:
`fn drawThePage[@runtime T: Draw](...)`

This is effectively a compiler hint, saying "I would like you to avoid specializing this function at compile-time". It's reminiscent of how the `inline` keyword is used in many languages.

The compiler would need to have a means of bridging code that _is_ specialized with code that is not, but let's avoid discussing this, because it will send us off-topic. More generally: let's avoid turning this thread into a discussion about runtime generics. (If anyone wants to talk about this further, e.g. reply to anything I have said here, we can create a separate Github Discussions thread.)

To summarise: I believe that I have demonstrated that Rust-style "trait objects" (which are constructed _manually_ by the programmer, by placing values into a `Box`) are undesirable. Instead, if runtime polymorphism is desired, it should be handled via the much broader notion of runtime generics. In other words, we should be thinking about how to handle parameters `[..]` at runtime.

> A final question is about the rebinding. In Rust that would be considered a downcast and could be fallible.

The rebinding is actually an _upcast_. `List[A | B | C]` is a more general type than `List[A | B]`. Thus, the cast is infallible.

More generally, union types can have subtypes and supertypes. This means we need to talk about [covariance and contravariance](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)) when discussing them.  The `List` type constructor is covariant with respect to its parameter.

> Some APIs use traits as a way to make opaque types so that they can avoid leaking implementation details to users (such as returning impl Iterator).

Opaque types are valuable, but they're orthogonal to trait objects, and to this proposal. Opaque types can be handled at compile-time, just as type parameters are.

**Comment by lsh** [comment]
> `List[A | B | C]` is a more general type than `List[A | B]`. Thus, the cast is infallible.

This seems to ignore the in memory representation of the data. In TypeScript this is fine because the types make no guarantee about memory representation, but Mojo _does_ make a guarantee. That means this either is fallible if the type is a superset (because the size might be larger), or it will have to go through and reallocate each item (which I would want called out in a much more apparent way).

**Comment by nmsmith** [comment]
The memory layout is determined at compile-time, when `drawThePage` is monomorphized. Specifically, it is determined at compile-time that `myItems` (the list passed into the function) will eventually hold elements of the type `Button | Image | MenuBar | SideBar`, thus the compiler will generate a _de-facto_ enum with 4 elements and a well-defined memory layout.

The "cast" is nothing more than a type annotation. It has no effect on the runtime behaviour of the program. As I stated in the original post, it may not even be required.

**Comment by lsh** [comment]
So wait, does `let items: List[T | MenuBar | SideBar] = items` allocate? Because before the types were `Image | Button`?
Let's say you had:
```
fn drawThePage[T: Draw](borrowed items: List[T]):
    let items: List[T | MenuBar | SideBar] = items

fn drawThePage2[T: Draw](borrowed items: List[T]):
    let items: List[T | MenuBar | Footer] = items
let myItems: List[Button | Image] = [Button(...), Image(...), Image(...)]
drawThePage(items)
drawThePage2(items)
```

What is the layout in memory?

**Comment by lsh** [comment]
Let's imagine for the sake of this exercise that `Footer` is significantly larger than every other type. Am I paying for that everywhere because I used it once?

**Comment by lsh** [comment]
Follow up: if I comment out that last line, is the codegen and memory footprint different because `items` never passes through `drawThePage2`?

**Comment by nmsmith** [comment]
Good question!

I shall assume you meant `inout` instead of `borrowed`, thereby allowing the functions to modify the collection. (Otherwise, no `MenuBar`s will ever be added.)

In this case, the compiler would look at all the places where `myItems` is used, and gather all of the variants that can possibly be added to the collection. The variants are: `Button | Image | MenuBar | SideBar | Footer`. It will then generate an enum with 5 variants and a well-defined memory layout. The very first items that are added to the list (the `Button` and the two `Image`s) will be stored in the list using this layout.

Basically, the compiler needs to look at all the future uses of each data structure to figure out what the memory layout should be.

> Let's imagine for the sake of this exercise that Footer is significantly larger than every other type. Am I paying for that everywhere.

Yes, that would be the case. But...
- If you have a large data type like this, you should be heap-allocating it to begin with.
- Maybe the compiler could decide to heap-allocate enum members (and in general, stack-allocated values) if they are unacceptably large. (It's an idea worth considering, at least.)

> if I comment out that last line, is the codegen and memory footprint different because items never passes through drawThePage2?

Yes 😃

**Comment by lsh** [comment]
I am asking in the `borrowed` case, actually, because even if it isn't used, it might still effect the type! Am I paying for a variant I might not use?

**Comment by lsh** [comment]
Also, I would consider it much harder to reason about memory with these semantics. Working with clear fixed sized structs was something I was looking forward to, and having the size of data suddenly shifting out from under me is nerve wracking.

**Comment by lsh** [comment]
I also definitely want heap allocations to be a decision I make and not the compiler.

**Comment by nmsmith** [comment]
> Am I paying for a variant I might not use?

If `drawThePage` doesn't have the authority to add extra variants to the List, then no, obviously the compiler wouldn't choose a memory layout that accommodates for those variants.

> I also definitely want heap allocations to be a decision I make

Sure. That was just a side-idea. It's not the design I'm pushing for.

**Comment by nmn** [comment]
`roc` does have anonymous union types that work in a very interesting way.

> List[A | B | C] is a more general type than List[A | B]. Thus, the cast is infallible.

This is only true for an immutable list. A mutable list would be invariant.

**Comment by hardskulls** [comment]
> `roc` does have anonymous union types that work in a very interesting way.
> 
> > List[A | B | C] is a more general type than List[A | B]. Thus, the cast is infallible.
> 
> This is only true for an immutable list. A mutable list would be invariant.

Oh, now I get it...
Thanks!

**Comment by nmsmith** [comment]
I've had an epiphany recently. We might be able to model unions and trait objects really cleanly using dependent types. For this approach to work, we'd need the ability to:
1. Store type-values at runtime.
2. Pass around traits as values.

A Rust-style "trait object" could be modelled via a definition such as the following:
```
struct Object[Trait: AnyTrait]:
    var Type: Trait  # store the type at runtime
    var value: Type
    fn __init__[Type: Trait](inout self, value: Type):
        self.Type = Type
        self.value = value
```

**Caveat:** The above definition implies that `value` would be stored inline. Given that its type is not known until runtime, this would make `Object` a "dynamically-sized type". This would prevent instances from being stored in an array, which is not ideal. In practice, we'd probably want to store `value` indirectly, and access it via a pointer.

This type could be used as follows (ignoring the sizing issue):
```
var items: Array[Object[Stringable]] = [Object(Int(5)), Object(String("hello"))]

for item in items:
    if item.Type is Int:
        # Now `item.value` can be treated as an `Int`
        item.value += 1
```

Now, a trait object is different from an enum, because the set of members of an enum is _closed_. To model an enum, we'd need a way to specify that a type can only take on a value from a fixed set. I'll use the syntax `var Type: {Type1, Type2}` for this.

To define a struct that models enums, we'll need it to accept a variadic number of types. The following definition seems plausible:

```
struct Enum[*Types: AnyType]:
    var Type: {*Types}
    var value: Type
    fn __init__[Type: {*Types}](inout self, value: Type):
        self.Type = Type
        self.value = value
```

This data type can be used in the exact same manner as the `Object` type:
```
var items: Array[Enum[Int, String]] = [Enum(Int(5)), Enum(String("hello"))]

for item in items:
    if item.Type is Int:
        # Now `item.value` can be treated as an `Int`
        item.value += 1
```

This approach could be really promising. The beautiful thing about this approach is that **it doesn't require enums/unions to be a _language feature_.** Instead, they can be defined as a library, built on top of dependent types.

**Comment by Mogball** [comment]
While I agree it would be really nice to implement these as purely library types, in practice I don't think we will without some compiler magic.

The first is the use of dynamic type values in type annotations. I think this is well-formed (dispatch dynamic queries into the runtime type value instead of statically), but I wouldn't put it on the critical path. Also it would have to be a boxed value, so `Pointer[None]` is probably fine.

The second is that the struct has to magically implement all the methods of the trait. Our metaprogramming is not good enough to express something like this. This is where compiler magic needs to come in for now. 

Also, I think existentials should be spelled with `Any[T]` and `object[T]` is slightly different in that it contains an `Any[T]` but dynamically dispatches any methods, like Python objects.

The last thing is that with a first class IR representation, we can optimize dynamic dispatch away trivially.

**Comment by nmsmith** [comment]
> the struct has to magically implement all the methods of the trait. Our metaprogramming is not good enough to express something like this. This is where compiler magic needs to come in for now.

The same problem occurs for `Reference`, right? e.g. for some type `T: Fooable`, it would be nice for a `List[Reference[T]]` to be passable to a function that expects `List[some Fooable]`.

I can envisage a solution that works for both `Reference` and `Any`. Both of these data types are just "proxies" for another variable, so it makes sense that they can "inherit" the traits of their target variable. (Except perhaps for some of the built-in traits, e.g. `Movable`, where inheritability doesn't make sense.)

We can possibly integrate this idea with the notion of "dereferencing" (`__refitem__`). For example, any struct that can be dereferenced could inherit the traits of its referent—assuming that it doesn't implement those traits itself.

> The first is the use of dynamic type values in type annotations. I think this is well-formed (dispatch dynamic queries into the runtime type value instead of statically).

For trait objects (`Object/Any`), yes, I'd imagine that would be the implementation strategy. But for the `Enum` type above, given that enums are limited to a statically-known set of alternatives, I'd argue that the type field should be encoded as an integer indicating which of the alternatives it is (for space reasons), and then:
- **Struct methods** can be dispatched on `value` by offering flow-sensitive narrowing of the type field (in the style of TypeScript), and then doing static dispatch on whatever the type has been narrowed down to within a particular branch.
- **Trait methods** can be dispatched on `value` in any context where the type has been narrowed to a point where all remaining alternatives implement the trait. To obtain a pointer to the actual table of trait methods, you'd need to add a bit of generated code that switches on the type field (the integer) and grabs a pointer to the correct table.

More generally, any value with type `in <iterable>` can be encoded as an integer.

**Comment by nmsmith** [comment]
I'm starting to suspect that storing types in runtime variables, e.g. `var Type: AnyType`, doesn't really make sense. In particular, while it might be possible to store a type in a local variable or a field, one certainly couldn't store a type in a generic collection, because every such collection is defined in terms of a **type variable** with a trait bound, e.g. `T: CollectionElement`, which implies that the collection can only store _instances_ of types. And a type cannot be an instance of a type, because we need types and their instances to be in separate [universes](https://en.wikipedia.org/wiki/Universe_%28mathematics%29).

However, there might be a way to encode types _indirectly_ at runtime, via zero-size objects and some kind of variant/enum feature. For example:
```
struct Type[Value: AnyType]:
    pass
var SomeType: Variant[Type[Int], Type[String]] = ...
```
(I'm not sure how `Variant` itself would be defined, though.)

**Comment by felix-andreas** [comment]
Another approach to union types - which also avoids Rust's problem of overbroad errors types (e.g. [any file system operation](https://doc.rust-lang.org/stable/std/fs/fn.read_to_string.html) might return an `io::Error` of kind [`AddrInUse`](std::io::ErrorKind)) - are [Roc's tags](https://www.roc-lang.org/tutorial#combining-open-unions). [(For more see this talk)](https://www.youtube.com/watch?v=7SidSvJcPd0)

It's essentially a middle ground between Typescript and Rust: The variants are still nominal (or at least they have names), but you don't have to declare them first and instead can just make them up on the fly and errors accumulate automatically (like TypeScript), but it is still straightforward to pattern-match on them (like Rust).

In TypeScript the way to usually discriminate between different variants of a union is to check if certain fields are present or not (which sometimes can be cumbersome and error-prone).

**Comment by bomzj** [comment]
Look at Haskell, F#, no need to reinvent or copy/paste bad habbits from Rust.

**Comment by hardskulls** [comment]
Good news everyone!

<img src="https://github.com/user-attachments/assets/45168497-32c8-4c47-8a36-f4899a0699e5" width="150">

I asked `Roc's` devs if it was possible to use `Roc's` tags in for a low-level language, and they confirmed that yeah, it's 100 percent possible, and their implementation is a efficient as that of `Rust`.
Turns out that no one else is using it not because it was impossible, but because they are the first to do that sort of experiment (which turned out to be successful).

[Link](https://roc.zulipchat.com/#narrow/stream/231634-beginners/topic/Roc's.20tags)

**Comment by nmsmith** [comment]
@hardskulls as Richard points out on that thread, Roc's tags are basically the same as OCaml's polymorphic variants. So they're not necessarily anything novel.

That said, polymorphic variants are a point in the design space worth considering.

**Comment by hardskulls** [comment]
> @hardskulls as Richard points out on that thread, Roc's tags are basically the same as OCaml's polymorphic variants. So they're not necessarily anything novel.
> 
> That said, polymorphic variants are a point in the design space worth considering.

I thought he meant it looks the same, but differs in implementation. 
I looked at OCaml's docs and it's mentioned that polymorphic variants do carry a little overhead. 
But Richard compares it to Rust in terms of performance and efficiency in his talks.

**Comment by JUkhan** [comment]
Bringing Rust like enum and match is awesome and smart plus.
please do not add a barrier by the name of **super set of python**  
as pre now we can import python module in mojo that's enough. 
We want to see the Mojo multiple times next level.

**Comment by jorenham** [comment]
Union types have a big problem, but it's a rather subtle one (and it isn't really well known in the Python typing community as far as I can tell).

Let's say we have two simple types, `MaybePos` and `MaybeNeg`. In Python they can be defined as follows:

```py
from typing import Literal

type MaybePos = Literal[0, +1]
type MaybeNeg = Literal[0, -1]
``` 

<details>
<summary>
So <code>MaybePos</code> can be either 0 or +1, and <code>MaybeNeg</code> can be either 0 or -1. The <i>order</i> (the number of values it can assume) of both `MaybePos` and `MaybeNeg` is therefore 2. 
</summary>

The *bottom type* (aka the *empty type*, `typing.Never`, or `typing.NoReturn`) has an order of `0`, which is the smallest order. The *top type* (aka the *unit type*, or `builtins.object`) has the maximum order, but I'm not able to write it down.

</details>

The union of `MaybePos` and `MaybeNeg` can be written as `MaybePos | MaybeNeg`. This is exactly equivalent to `Literal[0, +1, -1]`, so it has an order of 3. So after combining two types of order 2, we get a new type of order 3. That's less than 2+2=4, so we have lost some information. That means that the union type cannot be reversed. In practice that means that we can no longer distinguish between `MaybePos` and `MaybeNeg`. But other kinds of combinations like the tuple type (aka the *product type*), i.e. `tuple[MaybePos, MaybeNeg]`, don't have that problem. 

---

Luckily, there's a different "kind of union" that also doesn't have this problem: The sum type, aka *tagged-* or *disjoint union*, and it's what the cool kids are using (e.g. ocaml, scala3, f#, rust, etc.). Without going into too much detail; it doesn't lose any type-information, and "feels" very much like a (non-disjoint) union type. It's the typing analogue of *logical disjunction* in [intuitionistic logic](https://wikipedia.org/wiki/Intuitionistic_logic). See <https://wikipedia.org/wiki/Algebraic_data_type> for some unique use-cases, and <https://wikipedia.org/wiki/Tagged_union> for the implementation details.

So to summarize: Don't replicate Python's union type, sum types are way cooler.

**Comment by chriselrod** [comment]
Note that mojo has [variant](https://docs.modular.com/mojo/stdlib/utils/variant/Variant/)s, a sum type.

**Comment by segeljakt** [comment]
On the "Union types have a big problem":

This seems like more of a problem related to refinement types than union types? Union over refinement types is set union, which merges overlapping values. If you want to distinguish `MaybePos` and `MaybeNeg` you would just use a newtype wrapper, which is effectively what happens when you use enums. Enums force you to assign each member of the union a unique tag. In union types, the type itself is the tag. There is otherwise not much difference between them.

**Comment by jorenham** [comment]
> This seems like more of a problem related to refinement types than union types?

Tagged unions/sum types are not related to refinement types. Refinement types are a special case of intersection types.
